database/interface.py

```diff
@@ -142,13 +149,117 @@ def update_harvest_source(self, source_id, updates):
             self.db.rollback()
             return None
 
+    def clear_harvest_source(self, source_id):
+        """
+        Clear all datasets related to a harvest source in CKAN, and clean up the
+        harvest_record and harvest_record_error tables.
+        :param source_id: ID of the harvest source to clear
+        """
+
+        # delete all HarvestRecords and related HarvestRecordErrors
+        def _clear_harvest_records():
+                self.db.query(HarvestRecordError).filter(
+                    HarvestRecordError.harvest_record_id.in_(
+                        self.db.query(HarvestRecord.id).filter_by(harvest_source_id=source_id)
+                    )
+                ).delete(synchronize_session=False)
+                self.db.query(HarvestRecord).filter_by(harvest_source_id=source_id).delete()
+                self.db.commit()
+
+        source = self.db.get(HarvestSource, source_id)
+        if source is None:
+            return "Harvest source not found"
+
+        organization_id = source.organization_id
+
+        records = (
+            self.db.query(HarvestRecord)
+            .filter_by(harvest_source_id=source_id).all()
+        )
+
+        if not records:
+            return "Harvest source has no records to clear."
+
+        ckan_ids = [record.ckan_id for record in records if record.ckan_id is not None]
+        error_records = [record for record in records if record.status == 'error']
+        jobs_in_progress = self.get_all_harvest_jobs_by_filter(
+            {"harvest_source_id": source.id, "status": "in_progress"}
+        )
+
+        # Ensure no jobs are in progress
+        if jobs_in_progress:
+            return ("Error: A harvest job is currently in progress. "
+                    "Cannot clear datasets.")
+
+        # Ensure (error_records + ckan_ids) = total records
+        if len(error_records) + len(ckan_ids) != len(records):
```

Comment:
can you explain this check?

*also i think there's some issue with excess double quoting here in the return val.